Config

wolfeman67 opened the pull request.

wolfeman67 committed:
Added a DAL with two methods and a JSON config file
These are all subject to change though, especially the DAL.

wolfeman67 committed:
This is temporary. I worked on this from home and didn't want to lose what I typed.

wolfeman67 committed:
Temporary

wolfeman67 committed:
Created a configuration parser DAL and fixed a couple of issues

These issues being that the tests' channels still give off StasisStart
events after they have been hungup.  And that a hangup of an inbound,
through logic of the code causes a hangup request to happen again
(same thing with the dialed channels).  The tests have their own
configuration files that are test specific now, which makes it easier
than manually retyping the code in sla.js each time you want a different
number of endpoints.

wolfeman67 committed:
Fixed the README file to better reflect the changes of configuration use

wolfeman67 committed:
Had an unneccessary variable in sla.js

wolfeman67 committed:
Fixed misceallaneous programming errors.
Unneccessary config file under lib
Unneccessary variable inside sla.js
Ensured that inbound channel doesn't get asked to hang up a second time
Fixed a discrepancy in README.md

wolfeman67 committed:
Reworked how the trunk and sharedExtension are inputted
For now, the trunk and sharedExtension must be the same.  The
incoming data from the dialplan extension will determine both the
trunk and sharedExtension to use. More trunks will be supported later.
Fixed issues with the tests.

samuelg commented on app.js:
```diff
@@ -43,7 +44,8 @@ function clientLoaded (client) {
  */
 function errHandler(err) {
-  if(err.name === 'EarlyHangup' || err.name === 'HangupFailure') {
+  if(err.name === 'EarlyHangup' || err.name === 'HangupFailure' ||
```
That is already a bit unwieldy and will only grow, I would move that check to a function. fatalError() perhaps?

samuelg commented on app.js:
```diff
@@ -29,8 +29,9 @@ function clientLoaded (client) {
-      var bridgeName = event.args[0];
-      sla(client, channel, bridgeName)
+      var extension = event.args[0];
+      var confFile = process.argv[2];
```
What happens if the conf file is not provided by the CLI?

samuelg commented on res/config.json:
```diff
@@ -0,0 +1,19 @@
+  {
+    "42": {
+      "trunks": [
+        {"42": 
```
This is hard to read. I would look at prettifying the json file so it's in a common format.

samuelg commented on tests/testConfigs/multipleEndpoints.json:
```diff
@@ -0,0 +1,19 @@
+        {"42":                                                                   
+          {                                                                      
+            "stations": [                                                        
+            {"endpoint": "SIP/phone1"                                      
```
Look at formatting here as well.

samuelg commented on tests/testConfigs/noEndpoints.json:
```diff
@@ -0,0 +1,16 @@
+  {                                                                              
+    "42": {                                                             
+      "trunks": [                                                                
+        {"42":                                                                   
```
Look at formatting here as well.

samuelg commented on tests/testConfigs/singleEndpoint.json:
```diff
@@ -0,0 +1,17 @@
+  {                                                                              
+    "42": {                                                             
+      "trunks": [                                                                
+        {"42":                                                                   
```
Look at formatting here as well.

samuelg commented on lib/dal.js:
```diff
@@ -0,0 +1,86 @@
+  var defer = Q.defer();
+  var sharedExtension = {};
+  sharedExtension.pos = -1;
+  return getConfig(confFile).then(function(data) {
```
You are putting an implementation detail directly in the API you expose. If you need to add database support later, you'll have a lot more work to separate this out from the functionality of returning a shared extension object. Also, what if you wanted to support both config files and a database?

You should have an internal only API for getting the data that can be modified to support various back-ends instead of having this inside the API you expose.

samuelg commented on lib/dal.js:
```diff
@@ -0,0 +1,86 @@
+var getSharedExtension = function(confFile, name) {
+  var defer = Q.defer();
+  var sharedExtension = {};
+  sharedExtension.pos = -1;
```
You could also have a boolean called found to determine this since -1 is rather cryptic.

samuelg commented on lib/dal.js:
```diff
@@ -0,0 +1,86 @@
+      defer.resolve(trunk);
+    }
+  });
+  if(!trunk.pos) {
```
I would use a boolean here.

samuelg commented on lib/sla.js:
```diff
@@ -98,46 +99,102 @@ function addChannelsToBridge(inbound, dialed, bridge) {
+  }
+
+  function dialedChannelHungup(event) {
+    numToHangup -= 1;
```
Why not remove the dialed channel from the dialed array and checking the length of the array instead? You could also have a dialed object with an array of channels that get updated when hungup and a areAllHungup() method.

samuelg commented on lib/sla.js:
```diff
@@ -98,46 +99,102 @@ function addChannelsToBridge(inbound, dialed, bridge) {
+          dialed.push(client.Channel());
+        });
+        numToHangup = dialed.length;
+        dialed.forEach(function (channel, i, dialed) {
```
i is not a very descriptive variable name. At the very least use index.

samuelg commented on lib/sla.js:
```diff
@@ -98,46 +99,102 @@ function addChannelsToBridge(inbound, dialed, bridge) {
+        numToHangup = dialed.length;
+        dialed.forEach(function (channel, i, dialed) {
+          originate.push(Q.denodeify(channel.originate.bind(dialed[i])));
+          originate[i]({endpoint: stations[i].endpoint, app: 'sla',
```
Why keep the bound originate methods around?

samuelg commented on lib/sla.js:
```diff
@@ -98,46 +99,102 @@ function addChannelsToBridge(inbound, dialed, bridge) {
+    .then(function (stations) {
+      if (stations.length) {
+        stations.forEach(function() {
+          dialed.push(client.Channel());
```
If you create an object with a station and a dialed channel, you'll be able to just iterate over that and not use an index anymore.

samuelg commented on lib/sla.js:
```diff
@@ -98,46 +99,102 @@ function addChannelsToBridge(inbound, dialed, bridge) {
+        dialed.forEach(function (channel, i, dialed) {
+          originate.push(Q.denodeify(channel.originate.bind(dialed[i])));
+          originate[i]({endpoint: stations[i].endpoint, app: 'sla',
+            appArgs: 'dialed', timeout: 10})
```
This should be indented 2 more spaces.

samuelg commented on lib/sla.js:
```diff
@@ -98,46 +99,102 @@ function addChannelsToBridge(inbound, dialed, bridge) {
-  }
+function getTrunkData (configurationFile, extension) {
+  var defer = Q.defer();
+  dal.getSharedExtension(configurationFile, extension)
```
There's no need for a deferred here. Just return dal.getSharedExtension and within it either return a value on success or throw an error if not.

Deferred should only be created explicitly if you are creating a promise supported API or if you want the result of an event listener firing to resolve a promise. Since dal.getSharedExtension is already a promise supported API there's no reason to create a deferred yourself.

wolfeman67 committed:
Fixed miscealleanous errors. Restructured things.
Essentially made the program more modular in how it accessed data
and how it originated calls.  Made the JSON files "prettier"

matt-jordan commented on res/config.json:
```diff
@@ -0,0 +1,22 @@
+        "trunks": [
+          {
+            "42": {
+              "stations": [
```
While it may be better to deal with this when you start manipulating outbound dial, stations should not be conceptually a child of trunks. Trunks are a type of endpoint that participate in a shared extension; they are a separate concept from the stations that they serve.

wolfeman67 replied:
I have an idea of how to fix this, but I'll fix it in the next issue.

matt-jordan commented on lib/dal.js:
```diff
@@ -0,0 +1,76 @@
+
+var getSharedExtension = function(confFile, name) {
+  var defer = Q.defer();
+  var sharedExtension = {};
```
From a stylistic perspective, it is often nice to have a line break between your variable declarations and the code that acts on them.

``` javascript
var defer = Q.defer();
var sharedExtension = {};

sharedExtension.found = false;
...
```

matt-jordan commented on lib/dal.js:
```diff
@@ -0,0 +1,76 @@
+  sharedExtension.found = false;
+  return fetchJSON(confFile).then(function(data) {
+    data.sharedExtensions.forEach(function(extension, i) {
+      if(extension[name]) {
```
Space between if and conditional:

``` javascript
if (extension[name]) {
```

matt-jordan commented on app.js:
```diff
@@ -38,12 +42,25 @@ function clientLoaded (client) {
+ * @return {boolean} - if the error is fatal or not
+ */
+function isFatal(err) {
+  if(err.name === 'EarlyHangup' || err.anme === 'HangupFailure' ||
```
Space between if and conditional:

``` javascript
if (err.name ...
```

matt-jordan commented on app.js:
```diff
@@ -29,8 +33,8 @@ function clientLoaded (client) {
-      var bridgeName = event.args[0];
-      sla(client, channel, bridgeName)
+      var extension = event.args[0];
+      sla(client, channel, confFile, extension)
```
Your parameter ordering here is a bit odd:

1: The ARI client (okay)
2: The channel entering the shared extension (passed in from Stasis, okay)
3: The config file for the shared extension (but what shared extension?)
4: The actual shared extension (passed in from Stasis...)

From an ease of use perspective, you should order your parameters such that they flow logically. One ordering would be:

``` javascript
client, confFile, channel, extension
```

In other words: lead with the things that are unlikely to ever change (client, confFile) and follow with the more mutable pieces of data.

matt-jordan commented on app.js:
```diff
@@ -6,11 +6,15 @@ var sla = require('./lib/sla.js');
+    .catch(errHandler)
+    .done();
+} else {
+  console.error('No configuration file specified');
```
Is it fatal to not provide or find a configuration file? I would imagine this should actually bail out of the application.

matt-jordan commented on lib/sla.js:
```diff
@@ -98,65 +99,129 @@ function addChannelsToBridge(inbound, dialed, bridge) {
-      defer.resolve(line);
+  var dialed = [];
+
+  function dialedEnteredStasis(event, line) {
```
Document your function

matt-jordan commented on lib/sla.js:
```diff
@@ -98,65 +99,129 @@ function addChannelsToBridge(inbound, dialed, bridge) {
+    defer.resolve(line);
+  }
+
+  var allAreHungup = function(dialed) {
```
Document this function

matt-jordan commented on lib/sla.js:
```diff
@@ -98,65 +99,129 @@ function addChannelsToBridge(inbound, dialed, bridge) {
+   }
+  };
+
+  var findInDialed = function(channel) {
```
Document this function

matt-jordan commented on lib/sla.js:
```diff
@@ -98,65 +99,129 @@ function addChannelsToBridge(inbound, dialed, bridge) {
+    return position;
+  };
+
+  function dialedChannelHungup(event) {
```
Document this function

matt-jordan commented on lib/sla.js:
```diff
@@ -98,65 +99,129 @@ function addChannelsToBridge(inbound, dialed, bridge) {
+    }
+  }
+
+  function inboundHungup(event, line) {
```
Document this function

matt-jordan commented on lib/fetchJSON.js:
```diff
@@ -0,0 +1,15 @@
+var fs = require('fs');
+var Q = require('q');
+
+/** Reads from the configuration file and returns data
```
From a stylistic perspective, the description should go on the next line:

``` javascript
/**
 * Read from the configuration file and returns data
```

This finding applies to multiple comment blocks

wolfeman67 committed:
Addressed stylistic findings in my code.

wolfeman67 committed:
Added spaces in if statements

wolfeman67 committed:
Realized that I forgot to change the testFile. Now fixed.

wolfeman67 committed:
Ensured that noEndpoints configuration really had no endpoints

wolfeman67 committed:
Forgot an additonal formatting space in test file.

samuelg commented on app.js:
```diff
@@ -38,12 +42,25 @@ function clientLoaded (client) {
  */
 function errHandler(err) {
-  if(err.name === 'EarlyHangup' || err.name === 'HangupFailure') {
+  if (isFatal(err)) {
```
It looks like you only log the error if it is fatal, but blow up if it is not fatal.

samuelg commented on app.js:
```diff
@@ -38,12 +42,25 @@ function clientLoaded (client) {
+ * @return {boolean} - if the error is fatal or not
+ */
+function isFatal(err) {
+  if (err.name === 'EarlyHangup' || err.anme === 'HangupFailure' ||
```
Since it evaluates to a boolean already, you can just use:

return err.name === 'EarlyHangup' || err.anme === 'HangupFailure' || ....;

samuelg commented on lib/sla.js:
```diff
@@ -98,75 +99,156 @@ function addChannelsToBridge(inbound, dialed, bridge) {
+   */
+
+  var allAreHungup = function(dialed) {
+   if (dialed.length) {
```
You can just use:

return dialed.length;

to get the same result. Although I don't think it's doing what you think it's doing. Any integer above 0 will evaluate to true, so if the array has channels that have not been hungup yet, it will evaluate that all channels have been hungup. You should do !dialed.length instead so it only evaluates to true once the array has had all its elements removed.

samuelg commented on lib/sla.js:
```diff
@@ -98,75 +99,156 @@ function addChannelsToBridge(inbound, dialed, bridge) {
+    });
+  })
+  .catch(function(invalid) {
+    if (invalid === 'Invalid specified extension: ' + extension) {
```
You should check invalid.name for the Exception reason.

samuelg commented on lib/sla.js:
```diff
@@ -183,37 +265,37 @@ function CustomError(name, message) {
+        });
+    })
+  .catch(function (err) {
+    if(err.message !== 'Inbound channel hungup') {
```
You should use a custom error and check err.name instead. Error messages can change, error names should not.

wolfeman67 committed:
Fixed my conditional statements
There were a couple that were returning the exact opposite intended
result, and there was one that was checking incorrect properties.
Also made it so that the application doesn't check err messages
but err names.

wolfeman67 committed:
Made CustomError modular and some other things
Made the originateChannel function give off a CustomError instead of
resolving with a string.  Also made the app.js file check for the
correct non-fatal CustomErrors.

samuelg commented:
Looks good. Make sure to squash into a single commit before merging.
